pkg/network/usm/utils/file_registry.go

```diff
@@ -40,170 +81,160 @@ func NewFileRegistry() *FileRegistry {
+			fileUnregisterErrors:         metricGroup.NewCounter("unregister_errors"),
+			fileUnregisterFailedCB:       metricGroup.NewCounter("unregister_failed_cb"),
+			fileUnregisterPathIDNotFound: metricGroup.NewCounter("unregister_path_id_not_found"),
 		},
 	}
 }
 
-// Register a ELF library root/libPath as be used by the pid
-// Only one registration will be done per ELF (system wide)
-func (r *FileRegistry) Register(root, libPath string, pid uint32, activate activationCB, deactivate deactivationCB) {
-	hostLibPath := root + libPath
-	pathID, err := NewPathIdentifier(hostLibPath)
+// Register inserts or updates a new file registration within to the `FileRegistry`;
+//
+// If no current registration exists for the given `PathIdentifier`, we execute
+// its *activation* callback. Otherwise, we increment the reference counter for
+// the existing registration if and only if `pid` is new;
+func (r *FileRegistry) Register(namespacedPath string, pid uint32, activationCB, deactivationCB callback) {
+	if activationCB == nil && deactivationCB == nil {
+		// if *both* callbacks are nil, there is no point in proceeding
+		return
+	}
+
+	path, err := NewFilePath(r.procRoot, namespacedPath, pid)
 	if err != nil {
-		// short living process can hit here
-		// as we receive the openat() syscall info after receiving the EXIT netlink process
-		if log.ShouldLog(seelog.TraceLvl) {
-			log.Tracef("can't create path identifier %s", err)
-		}
 		return
 	}
 
+	pathID := path.ID
 	r.m.Lock()
 	defer r.m.Unlock()
+	if r.stopped {
+		return
+	}
+
 	if _, found := r.blocklistByID[pathID]; found {
-		r.telemetry.libBlocked.Add(1)
+		r.telemetry.fileBlocked.Add(1)
 		return
 	}
 
 	if reg, found := r.byID[pathID]; found {
 		if _, found := r.byPID[pid][pathID]; !found {
 			reg.uniqueProcessesCount.Inc()
-			// Can happen if a new process opens the same so.
+			// can happen if a new process opens an already active file
 			if len(r.byPID[pid]) == 0 {
 				r.byPID[pid] = pathIdentifierSet{}
 			}
 			r.byPID[pid][pathID] = struct{}{}
 		}
-		r.telemetry.libAlreadyRegistered.Add(1)
+		r.telemetry.fileAlreadyRegistered.Add(1)
 		return
 	}
 
-	if err := activate(pathID, root, libPath); err != nil {
-		log.Debugf("error registering library (adding to blocklist) %s path %s by pid %d : %s", pathID.String(), hostLibPath, pid, err)
-		// we are calling UnregisterCB here as some uprobes could be already attached, UnregisterCB cleanup those entries
-		if deactivate != nil {
-			if err := deactivate(pathID); err != nil {
-				log.Debugf("UnregisterCB library %s path %s : %s", pathID.String(), hostLibPath, err)
-			}
-		}
-		// save sentinel value, so we don't attempt to re-register shared
-		// libraries that are problematic for some reason
+	if err := activationCB(path); err != nil {
+		// we are calling `deactivationCB` here as some uprobes could be already attached
+		_ = deactivationCB(FilePath{ID: pathID})
```

Comment:
there is not guarantee in your code that either `activationCB` or `deactivationCB` are not null (only that BOTH are not null)